### PR TITLE
2269 publish events from smbexec exploiter

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #2269.

Updates smbexec exploiter to publish exploitation and propagation events.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [x] ~~Was the documentation framework updated to reflect the changes?~~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] ~~Added relevant unit tests?~~
* [x] ~~Have you successfully tested your changes locally? Elaborate:~~
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/infection_monkey/exploit/smbexec.py [hunk 1]
```diff
@@ -30,14 +50,76 @@ class SMBExploiter(HostExploiter):
...
+        )
+
+    def _PROPAGATION_TAGS(self) -> Tuple[str, ...]:
+        return (
+            SMBEXEC_EXPLOITER_TAG,
+            T1021_ATTACK_TECHNIQUE_TAG,
+            T1105_ATTACK_TECHNIQUE_TAG,
+            T1210_ATTACK_TECHNIQUE_TAG,
+        )
+
```
Comment (shreyamalviya) [inline monkey/infection_monkey/exploit/smbexec.py:68]:
```suggestion
    _EXPLOITER_TAGS = (
        SMBEXEC_EXPLOITER_TAG,
        T1021_ATTACK_TECHNIQUE_TAG,
        T1110_ATTACK_TECHNIQUE_TAG,
        T1210_ATTACK_TECHNIQUE_TAG,
    )
    _PROPAGATION_TAGS = (
        SMBEXEC_EXPLOITER_TAG,
        T1021_ATTACK_TECHNIQUE_TAG,
        T1105_ATTACK_TECHNIQUE_TAG,
        T1210_ATTACK_TECHNIQUE_TAG,
    )

```

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

**Review by shreyamalviya — APPROVED** [review]

File: monkey/infection_monkey/exploit/smbexec.py [hunk 2]
```diff
@@ -64,29 +142,27 @@ def _exploit_host(self):
...
+                    self._publish_exploitation_event(timestamp, False, error_message=message)
 
             except Exception as exc:
-                logger.error(
+                message = (
                     f"Error while trying to copy file using SMB to {self.host.ip_addr} with "
                     f"{creds_for_log}:{exc}"
                 )
+                logger.error(message)
+                self._publish_exploitation_event(timestamp, False, error_message=message)
```
Comment (ilija-lazoroski) [inline monkey/infection_monkey/exploit/smbexec.py:160]:
```suggestion
                error_message = (
                    f"Error while trying to copy file using SMB to {self.host.ip_addr} with "
                    f"{creds_for_log}:{exc}"
                )
                logger.error(error_message)
                self._publish_exploitation_event(timestamp, False, error_message=error_message)
```

File: monkey/infection_monkey/exploit/smbexec.py [hunk 3]
```diff
@@ -29,15 +49,73 @@ class SMBExploiter(HostExploiter):
...
+        )
+        return self.exploit_result
+
+    def _exploit(self, dest_path: PurePath) -> Tuple[Optional[str], SelectedCredentials, float]:
+        agent_binary = self.agent_binary_repository.get_agent_binary(self.host.os["type"])
+        creds = generate_brute_force_combinations(self.options["credentials"])
         for user, password, lm_hash, ntlm_hash in interruptible_iter(creds, self.interrupt):
             creds_for_log = get_credential_string([user, password, lm_hash, ntlm_hash])
 
+            timestamp = time()
```
Comment (ilija-lazoroski) [inline monkey/infection_monkey/exploit/smbexec.py:118]: With this approach, both exploitation and propagation will have the same timestamp which is wrong.

**Review by ilija-lazoroski — CHANGES_REQUESTED** [review]
Timestamps

File: monkey/infection_monkey/exploit/smbexec.py (hunk 3 above)
Comment (cakekoa) [reply monkey/infection_monkey/exploit/smbexec.py:118]: As discussed, we will use the timestamp from when we copy the file for propagation. We will re-evaluate if/when we add events for copying the file.

File: monkey/infection_monkey/exploit/smbexec.py [hunk 4]
```diff
@@ -29,15 +50,74 @@ class SMBExploiter(HostExploiter):
...
+
+        scmr_rpc = self._get_rpc_connection(creds)
+
+        if not scmr_rpc:
+            msg = "Failed to establish an RPC connection over SMB"
 
+            logger.warning(msg)
+            self.exploit_result.error_message = msg
+
+            return self.exploit_result
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/smbexec.py:92]: Publish propagation failed

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/exploit/smbexec.py (hunk 4 above)
Comment (ilija-lazoroski) [reply monkey/infection_monkey/exploit/smbexec.py:92]: on it

**Pull request merged**